system/src/system_ledger_internal.cpp

```diff
@@ -0,0 +1,880 @@
+    char path[MAX_PATH_LEN + 1];
+    if (stagedSeqNum > 0) {
+        // The most recent data is staged
+        CHECK(getStagedFilePath(path, sizeof(path), ledger->name(), stagedSeqNum));
+        staged_ = true;
+    } else {
+        CHECK(getCurrentFilePath(path, sizeof(path), ledger->name()));
+    }
+    // Open the respective file
+    FsLock fs;
+    CHECK_FS(lfs_file_open(fs.instance(), &file_, path, LFS_O_RDONLY));
+    NAMED_SCOPE_GUARD(closeFileGuard, {
+        int r = closeFile(fs.instance(), &file_);
+        if (r < 0) {
+            LOG(ERROR, "Error while closing file: %d", r);
+        }
+    });
+    // Read the header. The header data has already been validated at this point
+    CHECK(readDataHeader(fs.instance(), &file_, &dataSize_));
+    ledger_ = ledger;
+    open_ = true;
+    closeFileGuard.dismiss();
+    return 0;
+}
+
+int LedgerReader::read(char* data, size_t size) {
+    if (!open_) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    size_t bytesToRead = std::min(size, dataSize_ - dataOffs_);
+    if (size > 0 && bytesToRead == 0) {
+        return SYSTEM_ERROR_END_OF_STREAM;
+    }
+    FsLock fs;
+    size_t n = CHECK_FS(lfs_file_read(fs.instance(), &file_, data, bytesToRead));
+    if (n != bytesToRead) {
+        return SYSTEM_ERROR_LEDGER_INVALID_FORMAT; // Unexpected end of file
+    }
+    dataOffs_ += n;
+    return n;
+}
+
+int LedgerReader::close(bool /* discard */) {
+    if (!open_) {
+        return 0;
+    }
+    // Consider the reader closed regardless if any of the operations below fails
+    open_ = false;
+    FsLock fs;
+    int result = closeFile(fs.instance(), &file_);
+    if (result < 0) {
+        LOG(ERROR, "Error while closing file: %d", result);
+    }
+    // Let the ledger flush any data
+    int r = ledger_->readerClosed(staged_);
+    if (r < 0) {
+        LOG(ERROR, "Failed to flush ledger data: %d", r);
+        return r;
+    }
+    return result;
+}
+
+int LedgerWriter::init(LedgerWriteSource src, int tempSeqNum, Ledger* ledger) {
+    // Create a temporary file
+    char path[MAX_PATH_LEN + 1];
+    CHECK(getTempFilePath(path, sizeof(path), ledger->name(), tempSeqNum));
+    FsLock fs;
+    CHECK_FS(lfs_file_open(fs.instance(), &file_, path, LFS_O_WRONLY | LFS_O_CREAT | LFS_O_EXCL));
+    NAMED_SCOPE_GUARD(closeFileGuard, {
+        int r = closeFile(fs.instance(), &file_);
+        if (r < 0) {
+            LOG(ERROR, "Error while closing file: %d", r);
+        }
+    });
+    // Reserve space for the header
+    CHECK_FS(lfs_file_seek(fs.instance(), &file_, sizeof(LedgerDataHeader), LFS_SEEK_SET));
+    ledger_ = ledger;
+    src_ = src;
+    tempSeqNum_ = tempSeqNum;
+    open_ = true;
+    closeFileGuard.dismiss();
+    return 0;
+}
+
+int LedgerWriter::write(const char* data, size_t size) {
+    if (!open_) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    if (src_ == LedgerWriteSource::USER && dataSize_ + size > LEDGER_MAX_DATA_SIZE) {
```

Comment:
If this happens, the ledger write will terminate with an error but `setLedgerData` doesn't set the discard flag on `ledger_close`. Will temporary ledger files accumulate or are they cleaned up some other way?

Reply:
The destructor of `LedgerWriter` discards the data.

Reply:
Hm, actually the app API doesn't close the stream with the discard flag set if writing fails. Good catch!

Reply:
Specifically, the scope guard that gets executed when a write error occurs should be closing the stream with `LEDGER_STREAM_CLOSE_DISCARD`: https://github.com/particle-iot/device-os/blob/feature/ledger/wiring/src/spark_wiring_ledger.cpp#L147-L149

Reply:
Fixed.